File: ts/project/.sdk/tm/c/feature/audit.c
// Audit trail (mirrors feature/audit.rs). Emits a structured record for every
// operation — who (actor), what (entity + op), the outcome, and a correlation
// id — suitable for compliance logging. Records accumulate on the feature
// (bounded by `max`, default 1000) and, when a `sink` callback is supplied,
// are also pushed to it. The actor is the per-call ctrl actor, falling back to
// the options `actor`, then "anonymous". Each operation is audited exactly once
// (the per-context marker in ctx.out prevents a PreDone + PreUnexpected
// double-log). Timestamps use the injectable `now` clock so tests stay
// deterministic.

#include "sdk.h"

#include <stdlib.h>
#include <string.h>

#define AUDIT_SEEN_KEY "audit_seen"

typedef struct {
  Feature base;
  char* name;
  bool active;
  voxgig_value* add_opts;
  voxgig_value* options;
  int64_t seq;

  // Activity tracking (mirrors the ts client._audit record).
  voxgig_value* records; // List of Value
} AuditFeature;

static void audit_emit(AuditFeature* af, Context* ctx, const char* outcome) {
  if (!af->active) {
    return;
  }

  // One record per operation (PreDone + a following PreUnexpected on a failure
  // must not double-log).
  voxgig_value* seen = ctx_out_extra_get(ctx, AUDIT_SEEN_KEY);
  if (voxgig_is_bool(seen) && voxgig_as_bool(seen)) {
    return;
  }
  ctx_out_extra_set(ctx, AUDIT_SEEN_KEY, v_bool(true));

  af->seq += 1;

  const char* actor = "anonymous";
  const char* opt_actor = fopt_str(af->options, "actor", "");
  if (opt_actor[0] != '\0') {
    actor = opt_actor;
  }
  if (ctx->ctrl && ctx->ctrl->actor && ctx->ctrl->actor[0] != '\0') {
    actor = ctx->ctrl->actor;
  }

  const char* entity = ctx->op->entity;
  const char* opname = ctx->op->name;

  voxgig_value* record = v_map();
  setp(record, "seq", v_num((double)af->seq));
  setp(record, "ts", v_num((double)fopt_now_call(af->options)));
  setp(record, "actor", v_str(actor));
  setp(record, "entity", v_str(entity));
  setp(record, "op", v_str(opname));
  setp(record, "outcome", v_str(outcome));
  setp(record, "correlationId", v_str(ctx->id));
  if (ctx->result) {
    setp(record, "status", v_num((double)ctx->result->status));
  }

  voxgig_list_push(voxgig_as_list(af->records), voxgig_retain(record));
  int64_t max = fopt_int(af->options, "max", 1000);
  while (voxgig_list_len(voxgig_as_list(af->records)) > (size_t)max) {
    voxgig_list_erase(voxgig_as_list(af->records), 0);
  }

  voxgig_value* sink = getp(af->options, "sink");
  if (voxgig_is_func(sink)) {
    call_vfn(sink, record);
  }
}

static void audit_pre_done(AuditFeature* af, Context* ctx) {
  // Outcome reflects the actual result; a non-2xx reaches PreDone before the
  // pipeline errors.
  bool ok = ctx->result ? (ctx->result->ok && ctx->result->err == NULL) : false;
  audit_emit(af, ctx, ok ? "ok" : "error");
}

static void audit_pre_unexpected(AuditFeature* af, Context* ctx) {
  audit_emit(af, ctx, "error");
}

static const char* audit_name(Feature* f) { return ((AuditFeature*)f)->name; }
static bool audit_active(Feature* f) { return ((AuditFeature*)f)->active; }
static voxgig_value* audit_add_options(Feature* f) { return ((AuditFeature*)f)->add_opts; }

static void audit_init(Feature* f, Context* ctx, voxgig_value* options) {
  (void)ctx;
  AuditFeature* af = (AuditFeature*)f;
  af->options = options;
  af->active = fopt_bool(options, "active", false);
  af->seq = 0;
}

static void audit_hook(Feature* f, const char* name, Context* ctx) {
  AuditFeature* af = (AuditFeature*)f;
  if (strcmp(name, "PreDone") == 0) {
    audit_pre_done(af, ctx);
  } else if (strcmp(name, "PreUnexpected") == 0) {
    audit_pre_unexpected(af, ctx);
  }
}

static voxgig_value* audit_track(Feature* f) {
  AuditFeature* af = (AuditFeature*)f;
  int64_t n = (int64_t)voxgig_list_len(voxgig_as_list(af->records));
  return cmap(2, "records", v_num((double)n), "seq", v_num((double)af->seq));
}

static const FeatureVT AUDIT_VT = {
  audit_name, audit_active, audit_add_options, audit_init, audit_hook,
  audit_track,
};

Feature* feature_audit_new(void) {
  AuditFeature* af = (AuditFeature*)calloc(1, sizeof(AuditFeature));
  af->base.vt = &AUDIT_VT;
  af->name = strdup("audit");
  af->active = true;
  af->add_opts = NULL;
  af->options = voxgig_new_undef();
  af->seq = 0;
  af->records = voxgig_new_list();
  return (Feature*)af;
}

File: ts/project/.sdk/tm/elixir/lib/projectname/feature/audit.ex
# ProjectName SDK audit feature
#
# Audit trail. Emits a structured record for every operation — who (actor),
# what (entity + op), the outcome, and a correlation id — suitable for
# compliance logging. Records accumulate on `client._audit["records"]`
# (bounded by `max`, default 1000) and, when a `sink` callback is supplied,
# are also pushed to it (e.g. to forward to a SIEM). The actor is taken
# from a per-call `ctrl` actor, then options (`actor`), then 'anonymous'.
# Timestamps use the injectable `now` clock so tests stay deterministic.

defmodule ProjectName.Feature.Audit do
  alias Voxgig.Struct, as: S
  alias ProjectName.Feature

  def new do
    f = Feature.base("audit")
    Feature.install(f, "init", fn ctx, opts -> init(f, ctx, opts) end)
    Feature.install(f, "PreDone", fn ctx -> pre_done(f, ctx) end)
    Feature.install(f, "PreUnexpected", fn ctx -> pre_unexpected(f, ctx) end)
    f
  end

  def init(f, ctx, options) do
    Feature.init_common(f, ctx, options)
    S.setprop(f, "seq", 0)

    client = S.getprop(f, "client")

    if S.getprop(client, "_audit") == nil do
      S.setprop(client, "_audit", S.jm(["records", S.jt([])]))
    end

    nil
  end

  # Outcome reflects the actual result; a non-2xx reaches PreDone before the
  # pipeline raises.
  defp pre_done(f, ctx) do
    if Feature.active?(f) do
      result = S.getprop(ctx, "result")
      ok = result != nil and S.getprop(result, "ok") != false and S.getprop(result, "err") == nil
      emit(f, ctx, if(ok, do: "ok", else: "error"))
    end

    nil
  end

  defp pre_unexpected(f, ctx) do
    if Feature.active?(f), do: emit(f, ctx, "error")
    nil
  end

  defp emit(f, ctx, outcome) do
    # One record per operation (PreDone + a following PreUnexpected on a
    # non-2xx must not double-log).
    if S.getprop(ctx, "_audit_seen") == true do
      nil
    else
      S.setprop(ctx, "_audit_seen", true)

      seq = S.getprop(f, "seq") + 1
      S.setprop(f, "seq", seq)

      opts = Feature.opts(f)

      ctrl = S.getprop(ctx, "ctrl")
      actor = if ctrl != nil, do: S.getprop(ctrl, "actor"), else: nil
      actor = if actor == nil, do: S.getprop(opts, "actor"), else: actor
      actor = if actor == nil, do: "anonymous", else: actor

      op = S.getprop(ctx, "op")
      entity_v = or_underscore(if op != nil, do: S.getprop(op, "entity"), else: nil)
      opname = or_underscore(if op != nil, do: S.getprop(op, "name"), else: nil)

      result = S.getprop(ctx, "result")
      status = if result != nil, do: S.getprop(result, "status"), else: nil

      record =
        S.jm([
          "seq", seq,
          "ts", Feature.now(f),
          "actor", actor,
          "entity", entity_v,
          "op", opname,
          "outcome", outcome,
          "status", status,
          "correlationId", S.getprop(ctx, "id")
        ])

      client = S.getprop(f, "client")
      records = S.getprop(S.getprop(client, "_audit"), "records")
      Feature.list_push(records, record)

      mxv = S.getprop(opts, "max")
      mx = if mxv == nil, do: 1000, else: mxv
      trim(records, mx)

      sink = S.getprop(opts, "sink")

      if S.isfunc(sink) do
        try do
          sink.(record)
        rescue
          _ -> nil
        catch
          _, _ -> nil
        end
      end

      nil
    end
  end

  # Mirrors Python `x or "_"`: empty string / nil / false collapse to "_".
  defp or_underscore(v) do
    if v == nil or v == false or v == "", do: "_", else: v
  end

  defp trim(records, mx) do
    if S.size(records) > mx do
      S.delprop(records, 0)
      trim(records, mx)
    end
  end
end
